test/integration_test.go

```diff
@@ -1976,26 +1982,87 @@ func (ts *IntegrationTestSuite) TestMaxConcurrentSessionExecutionSize() {
 	// schedule-to-start of the session creation worker)
 	err := ts.executeWorkflow("test-max-concurrent-session-execution-size", ts.workflows.AdvancedSession, nil,
 		&AdvancedSessionParams{SessionCount: 4, SessionCreationTimeout: 2 * time.Second})
-	// Confirm it failed on the 4th session
+	// Confirm it failed on the 4th session because it took to long to create
 	ts.Error(err)
 	ts.Truef(strings.Contains(err.Error(), "failed creating session #4"), "wrong error, got: %v", err)
+	ts.Truef(strings.Contains(err.Error(), "activity ScheduleToStart timeout"), "wrong error, got: %v", err)
+}
+
+func (ts *IntegrationTestSuite) TestMaxConcurrentSessionExecutionSizeWithRecreationAndWait() {
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+	defer cancel()
+	var manualCancel context.CancelFunc
+	ts.activities.manualStopContext, manualCancel = context.WithCancel(ctx)
+	// Create 2 workflows each wanting to create 2 sessions (second session on
+	// each is recreation to ensure counter works). This will hang with one
+	// creating 2 and another creating 1 and waiting. Then when we send the signal
+	// that was done creating sessions, they will complete theirs allowing the
+	// other pending creation to complete.
+	run1, err := ts.client.ExecuteWorkflow(ctx,
+		ts.startWorkflowOptions("test-max-concurrent-session-execution-size-recreate-1"),
+		ts.workflows.AdvancedSession, &AdvancedSessionParams{
+			SessionCount:           2,
+			SessionCreationTimeout: 40 * time.Second,
+			RecreateAtIndex:        1,
+		})
+	ts.NoError(err)
+	// Wait until sessions created
+	ts.waitForQueryTrue(run1, "sessions-created-equals", 2)
+
+	// Now create second and wait until create pending after 1
+	run2, err := ts.client.ExecuteWorkflow(ctx,
+		ts.startWorkflowOptions("test-max-concurrent-session-execution-size-recreate-2"),
+		ts.workflows.AdvancedSession, &AdvancedSessionParams{
+			SessionCount:           2,
+			SessionCreationTimeout: 40 * time.Second,
+			RecreateAtIndex:        1,
+		})
+	ts.NoError(err)
+	// Wait until sessions created
+	ts.waitForQueryTrue(run2, "sessions-created-equals-and-pending", 1)
+
+	// Now let the activities complete which lets run1 complete and free up
+	// sessions for run2
+	manualCancel()
+	ts.NoError(run2.Get(ctx, nil))
```

Comment:
Is it worth checking that run1 completes?
We usually verify all workflows finish in Core/TS tests for good measure.

Reply:
I will add the check.

Reply:
Done